### ISSUE TITLE
Handle the new calling convention that julia emits for cheaper ptls

### ISSUE THREAD

**Issue by motabbara** [EnzymeAD/Enzyme.jl#909]
Looks like something has gone quite wrong here with the presence of swiftcc and swiftself as a first argument. 

MWE on Julia 1.10 7b565e3d8dc. toml/manifest attached

```
using Enzyme

function foo(a::Float32)
    a*40.0
end

autodiff(Reverse, foo, Active, Active(3.0f0))

```

---

mod = ; ModuleID = 'start'
source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

; Function Attrs: alwaysinline returns_twice
define internal swiftcc double @julia_foo_5929_inner({}*** nonnull swiftself %0, float %1) #0 !dbg !4 {
top:
  %2 = call {}*** @julia.get_pgcstack()
  %3 = bitcast {}*** %2 to {}**
  %current_task = getelementptr inbounds {}*, {}** %3, i64 -14
  %4 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %4, i64 15
  %5 = bitcast {}*** %2 to {}**
  %current_task1 = getelementptr inbounds {}*, {}** %5, i64 -14
  %ptls_field = getelementptr inbounds {}*, {}** %current_task1, i64 16
  %ptls_load = load {}*, {}** %ptls_field, align 8, !tbaa !8
  %ptls = bitcast {}* %ptls_load to {}**
  %6 = bitcast {}** %ptls to i64**
  %7 = getelementptr inbounds i64*, i64** %6, i64 2
  %safepoint = load i64*, i64** %7, align 8, !tbaa !12, !invariant.load !7
  fence sIncorrect number of arguments passed to called function!
  %2 = call swiftcc double @julia_foo_5929_inner({}*** %1), !dbg !31
LLVM.API.LLVMVerifyFunction(wrapper_f, LLVM.API.LLVMPrintMessageAction) = 1
wrapper_f = ; Function Attrs: returns_twice
define double @julia_foo_5929_inner.1({}** %0) #4 !dbg !4 {
entry:
  %1 = alloca {}**, align 8, !dbg !31
  store {}** %0, {}*** %1, align 8, !dbg !31
  %2 = call swiftcc double @julia_foo_5929_inner({}*** %1), !dbg !31
  ret double %2, !dbg !31
}

(parmsRemoved, retRemoved, prargs) = (UInt64[], false, Any[(cc = GPUCompiler.GHOST, typ = typeof(foo), arg_i = 1), (cc = GPUCompiler.BITS_REF, typ = Float32, arg_i = 2, codegen = (typ = {}***, i = 1))])
mod = ; ModuleID = 'start'
source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

; Function Attrs: alwaysinline returns_twice
define internal swiftcc double @julia_foo_5931_inner({}*** nonnull swiftself %0, float %1) #0 !dbg !4 {
top:
  %2 = call {}*** @julia.get_pgcstack()
  %3 = bitcast {}*** %2 to {}**
  %current_task = getelementptr inbounds {}*, {}** %3, i64 -14
  %4 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %4, i64 15
  %5 = bitcast {}*** %2 to {}**
  %current_task1 = getelementptr inbounds {}*, {}** %5, i64 -14
  %ptls_field = getelementptr inbounds {}*, {}** %current_task1, i64 16
  %ptls_load = load {}*, {}** %ptls_field, align 8, !tbaa !8
  %ptls = bitcast {}* %ptls_load to {}**
  %6 = bitcast {}** %ptls to i64**
  %7 = getelementptr inbounds i64*, i64** %6, i64 2
  %safepoint = load i64*, i64** %7, align 8, !tbaa !12, !invariant.load !7
  fence sIncorrect number of arguments passed to called function!
  %2 = call swiftcc double @julia_foo_5931_inner({}*** %1), !dbg !31
LLVM.API.LLVMVerifyFunction(wrapper_f, LLVM.API.LLVMPrintMessageAction) = 1
wrapper_f = ; Function Attrs: returns_twice
define double @julia_foo_5931_inner.1({}** %0) #4 !dbg !4 {
entry:
  %1 = alloca {}**, align 8, !dbg !31
  store {}** %0, {}*** %1, align 8, !dbg !31
  %2 = call swiftcc double @julia_foo_5931_inner({}*** %1), !dbg !31
  ret double %2, !dbg !31
}

(parmsRemoved, retRemoved, prargs) = (UInt64[], false, Any[(cc = GPUCompiler.GHOST, typ = typeof(foo), arg_i = 1), (cc = GPUCompiler.BITS_REF, typ = Float32, arg_i = 2, codegen = (typ = {}***, i = 1))])
mod = ; ModuleID = 'start'
source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

; Function Attrs: alwaysinline returns_twice
define internal swiftcc double @julia_foo_5933_inner({}*** nonnull swiftself %0, float %1) #0 !dbg !4 {
top:
  %2 = call {}*** @julia.get_pgcstack()
  %3 = bitcast {}*** %2 to {}**
  %current_task = getelementptr inbounds {}*, {}** %3, i64 -14
  %4 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %4, i64 15
  %5 = bitcast {}*** %2 to {}**
  %current_task1 = getelementptr inbounds {}*, {}** %5, i64 -14
  %ptls_field = getelementptr inbounds {}*, {}** %current_task1, i64 16
  %ptls_load = load {}*, {}** %ptls_field, align 8, !tbaa !8
  %ptls = bitcast {}* %ptls_load to {}**
  %6 = bitcast {}** %ptls to i64**
  %7 = getelementptr inbounds i64*, i64** %6, i64 2
  %safepoint = load i64*, i64** %7, align 8, !tbaa !12, !invariant.load !7
  fence sIncorrect number of arguments passed to called function!
  %2 = call swiftcc double @julia_foo_5933_inner({}*** %1), !dbg !31
LLVM.API.LLVMVerifyFunction(wrapper_f, LLVM.API.LLVMPrintMessageAction) = 1
wrapper_f = ; Function Attrs: returns_twice
define double @julia_foo_5933_inner.1({}** %0) #4 !dbg !4 {
entry:
  %1 = alloca {}**, align 8, !dbg !31
  store {}** %0, {}*** %1, align 8, !dbg !31
  %2 = call swiftcc double @julia_foo_5933_inner({}*** %1), !dbg !31
  ret double %2, !dbg !31
}

(parmsRemoved, retRemoved, prargs) = (UInt64[], false, Any[(cc = GPUCompiler.GHOST, typ = typeof(foo), arg_i = 1), (cc = GPUCompiler.BITS_REF, typ = Float32, arg_i = 2, codegen = (typ = {}***, i = 1))])
mod = ; ModuleID = 'start'
source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

; Function Attrs: alwaysinline returns_twice
define internal swiftcc double @julia_foo_5937_inner({}*** nonnull swiftself %0, float %1) #0 !dbg !4 {
top:
  %2 = call {}*** @julia.get_pgcstack()
  %3 = bitcast {}*** %2 to {}**
  %current_task = getelementptr inbounds {}*, {}** %3, i64 -14
  %4 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %4, i64 15
  %5 = bitcast {}*** %2 to {}**
  %current_task1 = getelementptr inbounds {}*, {}** %5, i64 -14
  %ptls_field = getelementptr inbounds {}*, {}** %current_task1, i64 16
  %ptls_load = load {}*, {}** %ptls_field, align 8, !tbaa !8
  %ptls = bitcast {}* %ptls_load to {}**
  %6 = bitcast {}** %ptls to i64**
  %7 = getelementptr inbounds i64*, i64** %6, i64 2
  %safepoint = load i64*, i64** %7, align 8, !tbaa !12, !invariant.load !7
  fence sIncorrect number of arguments passed to called function!
  %2 = call swiftcc double @julia_foo_5937_inner({}*** %1), !dbg !31
LLVM.API.LLVMVerifyFunction(wrapper_f, LLVM.API.LLVMPrintMessageAction) = 1
wrapper_f = ; Function Attrs: returns_twice
define double @julia_foo_5937_inner.1({}** %0) #4 !dbg !4 {
entry:
  %1 = alloca {}**, align 8, !dbg !31
  store {}** %0, {}*** %1, align 8, !dbg !31
  %2 = call swiftcc double @julia_foo_5937_inner({}*** %1), !dbg !31
  ret double %2, !dbg !31
}

(parmsRemoved, retRemoved, prargs) = (UInt64[], false, Any[(cc = GPUCompiler.GHOST, typ = typeof(foo), arg_i = 1), (cc = GPUCompiler.BITS_REF, typ = Float32, arg_i = 2, codegen = (typ = {}***, i = 1))])
mod = ; ModuleID = 'start'
source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

; Function Attrs: alwaysinline returns_twice
define internal swiftcc double @julia_foo_5939_inner({}*** nonnull swiftself %0, float %1) #0 !dbg !4 {
top:
  %2 = call {}*** @julia.get_pgcstack()
  %3 = bitcast {}*** %2 to {}**
  %current_task = getelementptr inbounds {}*, {}** %3, i64 -14
  %4 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %4, i64 15
  %5 = bitcast {}*** %2 to {}**
  %current_task1 = getelementptr inbounds {}*, {}** %5, i64 -14
  %ptls_field = getelementptr inbounds {}*, {}** %current_task1, i64 16
  %ptls_load = load {}*, {}** %ptls_field, align 8, !tbaa !8
  %ptls = bitcast {}* %ptls_load to {}**
  %6 = bitcast {}** %ptls to i64**
  %7 = getelementptr inbounds i64*, i64** %6, i64 2
  %safepoint = load i64*, i64** %7, align 8, !tbaa !12, !invariant.load !7
  fence sIncorrect number of arguments passed to called function!
  %2 = call swiftcc double @julia_foo_5939_inner({}*** %1), !dbg !31
LLVM.API.LLVMVerifyFunction(wrapper_f, LLVM.API.LLVMPrintMessageAction) = 1
wrapper_f = ; Function Attrs: returns_twice
define double @julia_foo_5939_inner.1({}** %0) #4 !dbg !4 {
entry:
  %1 = alloca {}**, align 8, !dbg !31
  store {}** %0, {}*** %1, align 8, !dbg !31
  %2 = call swiftcc double @julia_foo_5939_inner({}*** %1), !dbg !31
  ret double %2, !dbg !31
}

(parmsRemoved, retRemoved, prargs) = (UInt64[], false, Any[(cc = GPUCompiler.GHOST, typ = typeof(foo), arg_i = 1), (cc = GPUCompiler.BITS_REF, typ = Float32, arg_i = 2, codegen = (typ = {}***, i = 1))])

[Manifest.txt](https://github.com/EnzymeAD/Enzyme.jl/files/11856365/Manifest.txt)
[Project.txt](https://github.com/EnzymeAD/Enzyme.jl/files/11856366/Project.txt)

**Comment by wsmoses** [comment]
Actually julia itself is now emitting swiftcc/swiftself: https://github.com/JuliaLang/julia/pull/50093

We haven't adapted to those nightly changes yet, though.

**Comment by motabbara** [comment]
Is that down to changing how reinsert_gcmarker! is implemented or is there more to it?

**Comment by wsmoses** [comment]
At least the issue you posted, I think comes down to how to correctly count/index the corresponding LLVM arguments for a julia method

**Comment by vchuravy** [comment]
GPUCompiler made the necessary change https://github.com/JuliaGPU/GPUCompiler.jl/pull/473/files

**Comment by motabbara** [comment]
Have a PR for this that I'm cleaning up. Updated GPUCompiler and LLVM and then tried to sort out all the ctx and lack of codegen_world_age.

**Comment by motabbara** [comment]
Fixed by #917
